### ISSUE TITLE
Editing bug: creates and preserves styles from h1-h6 while turning it into paragraph

### ISSUE THREAD

**Issue by ile** [quilljs/quill#4039]
I could not reproduce this on Codepen, but I could reproduce it on my local machine and [Quill.com](https://quilljs.com/).

1. Create some content with p and h1
![image](https://github.com/quilljs/quill/assets/433707/a61dd9e3-0813-4b2a-ab6e-76b936717f68)
![image](https://github.com/quilljs/quill/assets/433707/0456e6b5-a790-481b-9ce1-eb2b9cfd4989)
2. Go to the empty line
![image](https://github.com/quilljs/quill/assets/433707/de7ff351-e1f3-4c92-b096-27f3a49c0d70)
3. Hit del key (on PC)
![image](https://github.com/quilljs/quill/assets/433707/43f34380-1fd9-4d9f-986e-f4452b2878fa)
4. You should see that the line is turned into  p  span  ... resembling the original  h1
![image](https://github.com/quilljs/quill/assets/433707/4c289c99-9225-42e4-aa7b-b1422f06fcea)
![image](https://github.com/quilljs/quill/assets/433707/84e9420c-c329-4fcf-a31f-96ed0bdccc08)

**Expected behavior**:

p with no style

**Actual behavior**:

p  span with styles

**Platforms**:

Brave

**Version**:

'2.0.0-rc.2'

**Comment by ile** [comment]
The styles eventually creep into the h1-h6 too. 

Feels like this should never happen in an HTML editor unless it's broken in some severe manner. I 

![image](https://github.com/quilljs/quill/assets/433707/8bd52c61-4c5f-4dc3-b66f-49b4969e111c)

**Comment by luin** [comment]
Thanks for reporting! Can you reproduce this in Chrome? I can't reproduce it with Mac + Chrome 122.

**Comment by ile** [comment]
Yes, it does happen in the latest Chrome of Linux  and Ubuntu 23.


![image](https://github.com/quilljs/quill/assets/433707/f91bf138-5f25-426a-a942-99e3a9408648)
![image](https://github.com/quilljs/quill/assets/433707/bb990332-6c40-4be8-ba5e-1f4b243e89b4)

**Comment by ile** [comment]
Happens with the latest (rc-5) version on https://quilljs.com demo.

**Comment by ile** [comment]
My platform is Linux btw.

**Comment by ile** [comment]
Still happens with 2.0 on [quill.com. ](https://quilljs.com/).

**Comment by ile** [comment]
Do you think it will be possible to fix this? For me it looks like the most important thing to fix, otherwise the editor isn't really much different than a plain contenteditable.

**Comment by luin** [comment]
Sorry for the delay! I just started looking into this. I tried Chrome 120 on Window and Chrome 121 on Linux, but haven't reproduced the issue yet. Would be good if you could help me to reproduce it:

1. Can you go to https://w3c.github.io/uievents/tools/key-event-viewer.html and press the Del key that can reproduce the bug and see what's the `keyCode`/`key`/`code`?
2. Would you try running Chrome without any extensions enabled?

**Comment by ile** [comment]
Hey,

Good to hear from you! Here's a little video I made with Chromium 124 on Ubuntu 23.

All browsers I tried on Ubuntu (Chromium, Brave, Firefox) gave keycode 46 for the 'del'.

[quilljs.webm](https://github.com/quilljs/quill/assets/433707/1452eed2-0d1d-4ab6-a9e5-32cbad17c854)

**Comment by ile** [comment]
The bug doesn't happen there - the source code will have `<h1>` like it should after deleting that line with `del`.

Here https://quilljs.com/playground/snow it happens yet in another way: there will be `<p>` in the code, but no styles.

**Comment by luin** [comment]
> the source code will have <h1> like it should after deleting that line with del.

By `the source code`, do you mean the inner HTML in the `.ql-editor`?

Can you also try the [link](https://quilljs.com/playground/snow#codeN4Ig9AlgdgJgpgDwHQAsAuBbANiAXCAHhggDcACCGAXgB0Q5i0B7AJzoD4CxiT2ap+BAM4BjFhAAOaMkJYja4aPGQArIRy6jxU9iAA0i2IiRq8IEUyhDpARwCuELFjJUyUOAHcyARQdOAFADkAMQMEMwsgXpkwPxkZBhMMHZYcEK4MXHxZMxMWABGAIYsGQDaWdlkpcBkKHCF8CVVAIzRAEzRAGaFWEJwALpkAL79ehXZpYH5eTBRZIHhPRAic4F2RixY0HCBo+PxkxAYhQDmO9GBFvAAtPlYTCIA1rtjUJV7b8Ov8RJYhSJwFAzOBNQIAYSYGAkTD6ZEKbzgEmWSBRUSyaDqGDgGUCQigTA8czAYDIrHm+Ts+TuO34QwAlABufj8YlkeyOLBIR5wACe02KMCQ+SU0BOQhcMSGTKg7KcXN5-JYgoaMAAQiKoCd-LFPtyeRkACwANm+tXhMFSLH8LHhZ2iFigaEQaDpmU+8QdQjycCQ9y1gWNZtgqVmjPGGIgQlQ5tSABE4KkndbbXB7ZYnQgXdL4kNaWGBDK-JyWEwmGgkCqAKIkOCOgAykad7itgT1MAJUDm-jgrqo7DdHssXtSvqYWp9HhQyxQdLzICGQA) and see what get printed in the console?

**Comment by ile** [comment]
> By the source code, do you mean the inner HTML in the .ql-editor?

Yes, exactly. So it looks like the handler solves the problem.

![image](https://github.com/quilljs/quill/assets/433707/31c9976a-d6d2-4dbf-93e4-abbc4378eb30)

![image](https://github.com/quilljs/quill/assets/433707/66681794-6663-41d2-b01a-2f3f389d1932)

**Comment by luin** [comment]
Okay that sounds good! I will double check the key code in different keyboard layouts but in the meantime you can use the code here to fix the issue: [link](https://quilljs.com/playground/snow#codeN4Ig9AlgdgJgpgDwHQAsAuBbANiAXCAHhggDcACCGAXgB0Q5i0B7AJzoD4CxiT2ap+BAM4BjFhAAOaMkJYja4aPGQArIRy6jxU9iAA0i2IiRq8IEUyhDpARwCuELFjJUyUOAHcyARQdOAFADkAMQMEMwsgXpkwPxkZBhMMHZYcEK4MXHxZMxMWABGAIYsGQDaWdlkpcBkKHCF8CVVAIzRAEzRAGaFWEJwALpkAL79ehXZpYH5eTBRZIHhPRAic4F2RixY0HCBo+PxkxAYhQDmO9GBFvAAtPlYTCIA1rtjUJV7b8Ov8RJYhSJwFAzOBNQIAYSYGAkTD6ZEKbzgEmWSBRUSyaDqGDgGUCQigTA8czAYDIrHm+Ts+TuO34QwAlABufj8eyOLBIR5wACe02KMCQDRgACElNATv5Yp9OVyMgAWABs3zIFichQkfRgGTQLDscCVQjsnU6EAQGTAAD0kAB+AAkYCVKHhMFSLH8LHhZ2iFigaEQaDpmU+8QxECEqCdqQAInBUr63R69crLL6EP6mZ8hrTGSAhkA)

**Comment by janwirth** [comment]
Seems to be okay?

https://github.com/quilljs/quill/assets/7485959/1e4daa65-7bb9-46a9-81e5-6a916c2a8919
